src/background/feeder/dapp-session.feeder.ts

```diff
@@ -0,0 +1,91 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { ErrorWithConsoleLog } from '../../utils/error-with-console-log'
+import { IDappSessionMessage } from '../../utils/message/dapp-session/dapp-session.message'
+import { DirectMessageReq } from '../../utils/message/message-handler'
+import { DappSessionManager } from '../dapp-session.manager'
+
+export class DappSessionFeeder {
+  constructor(private manager: DappSessionManager) {
+    this.serveEvents()
+  }
+  serveEvents(): void {
+    console.log('Register DappSessionFeeder event listeners...')
+
+    chrome.runtime.onMessage.addListener((message, sender) => {
+      if (!this.isInternalMessage(sender)) return
+
+      if (this.isDappSessionMessage(message, 'registerDappSession')) {
+        this.manager.register(message.payload[0], {
+          tabId: this.senderTabId(sender),
+          frameId: this.senderFrameId(sender),
+          originContentRoot: this.senderContentOrigin(sender),
+          frameContentRoot: this.senderFrameOrigin(sender),
+        })
+      }
+    })
+  }
+
+  private senderTabId(sender: chrome.runtime.MessageSender) {
+    if (!sender.tab) throw new ErrorWithConsoleLog(`DappSessionFeeder: sender does not have "tab" property`, sender)
+
+    if (!sender.tab!.id) {
+      throw new ErrorWithConsoleLog(`DappSessionFeeder: sender does not have "tab.id" property`, sender)
+    }
+
+    return sender.tab.id
+  }
+
+  /** If context is not in iframe it returns back -1 */
+  private senderFrameId(sender: chrome.runtime.MessageSender) {
+    if (!sender.frameId) {
+      return -1
+    }
+
+    return sender.frameId
+  }
+
+  /** Gives back the original content reference of the sender */
+  private senderContentOrigin(sender: chrome.runtime.MessageSender) {
+    if (!sender.tab) throw new ErrorWithConsoleLog(`DappSessionFeeder: sender does not have "tab" property`, sender)
```

Comment:
I think the two checks can be merged as `if (!sender?.tab?.url)`

Reply:
yeah, you are right, I just wanted to give very punctual response, but as I'm also checking it back it is not neccessary.